Feature/87 create service tracker inside use tracked service call

pnoltes opened the pull request:
This PR adds support for creating a service tracking in a `celix_bundleContext_useTrackedService(s)*` calls. 

This is done by calling the use callbacks outside of the (read) lock. The prevent that the service tracker can be removed during a use callback, an atomic use count is introduced.

pnoltes committed:
gh-87: Add test to call useTrackedService on the event thread.

pnoltes committed:
gh-87: Support creating service tracker inside a useTrackedService call

PengZheng commented on libs/framework/src/bundle_context.c:
```diff
@@ -1058,11 +1082,10 @@ bool celix_bundleContext_useServiceWithId(
-    char filter[64];
-    snprintf(filter, 64, "(%s=%li)", CELIX_FRAMEWORK_SERVICE_ID, serviceId);
+    char filter[32]; //20 is max long length
+    (void)snprintf(filter, sizeof(filter), "(%s=%li)", CELIX_FRAMEWORK_SERVICE_ID, serviceId);
```
A long can take value `INT64_MIN`, which need 20 bytes(including `-`). The total length of the buffer needs to be 1 (`(`) + 10 (CELIX_FRAMEWORK_SERVICE_ID) + 1 (`=`) + 20 (`%li`) + 1 (`)`) + 1 ('\0') = 34 bytes. This modification will lead to gcc warning complaining string truncation.

PengZheng commented on libs/framework/src/bundle_context.c:
```diff
@@ -1396,40 +1420,46 @@ static celix_service_tracker_t* celix_bundleContext_findServiceTracker(celix_bun
+    if (trkEntry) {
+        // note use count is only increased inside a read (shared) lock and ensures that the trkEntry is not freed and
+        // the trkEntry->tracker is not destroyed until the use count drops back to 0.
+        (void)__atomic_fetch_add(&trkEntry->useCount, 1, __ATOMIC_RELAXED);
```
This use of `__ATOMIC_RELAXED` is indeed correct.

PengZheng reviewed (changes requested):
Nice API improvement. 
There are some minor issues.

PengZheng committed:
gh-87: Address review issues in gh-734.

1. Use acquire-release pair to synchronize the service user with the service stopper.
2. Reduce CPU usage when waiting for service user.
3. Fix gcc -Wstringop-truncation warning.

PengZheng reviewed (approved):
I tried to fix the above issues. Please have a look at them before actually merging them.

pnoltes commented on libs/framework/src/bundle_context.c:
```diff
@@ -839,7 +841,28 @@ static void celix_bundleContext_removeServiceTrackerTracker(void *data) {
+    // note that the use count cannot be increased anymore, because the tracker is removed from the map
+    struct timespec start = celix_gettime(CLOCK_MONOTONIC);
+    int logCount = 0;
+    while (__atomic_load_n(&trkEntry->useCount, __ATOMIC_RELAXED) > 0) {
```
Thanks. I assumed ordering was not relevant, because a count increased only happened in the shared lock. 
But indeed this check and thus the destroy of the tracking needs to be ordered after the use count decrease.
